File: LeetCode/iter1/c++/623.add-one-row-to-tree.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;


struct TreeNode {
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	TreeNode* addOneRow(TreeNode* root, int v, int d) {
		if (d == 1) {
			TreeNode* tmp = new TreeNode(v);
			tmp->left = root;
			root = tmp;
			return root;
		}

		queue<pair<TreeNode*, int>> treeNodeQ;
		treeNodeQ.push(pair<TreeNode*, int>(root, 1));

		//locate level
		vector<TreeNode*> parentsV;
		while (!treeNodeQ.empty()) {
			pair<TreeNode*, int> nodePair = treeNodeQ.front();
			treeNodeQ.pop();
			if (nodePair.second == d - 1) {
				parentsV.push_back(nodePair.first);
			}
			else {
				if (nodePair.first->left != NULL) {
					treeNodeQ.push(pair<TreeNode*, int>(nodePair.first->left, nodePair.second + 1));
				}
				if (nodePair.first->right != NULL) {
					treeNodeQ.push(pair<TreeNode*, int>(nodePair.first->right, nodePair.second + 1));
				}
			}
		}

		//change parent and children
		for (int i = 0; i < parentsV.size(); i++) {
			TreeNode* leftAddedNode = new TreeNode(v);
			leftAddedNode->left = parentsV[i]->left;
			parentsV[i]->left = leftAddedNode;

			TreeNode* rightAddedNode = new TreeNode(v);
			rightAddedNode->right = parentsV[i]->right;
			parentsV[i]->right = rightAddedNode;
		}

		return root;
	}
};

void testPrint(TreeNode* treeNode) {
	queue<pair<TreeNode*, int>> treeQ;
	treeQ.push(pair<TreeNode*, int>(treeNode, 1));
	int curLv = 1;

	while (!treeQ.empty()) {
		pair<TreeNode*, int>tmpPair = treeQ.front();
		treeQ.pop();

		if (curLv != tmpPair.second) {
			curLv = tmpPair.second;
			cout << endl;
		}

		cout << tmpPair.first->val << " ";
		if (tmpPair.first->left != NULL) {
			treeQ.push(pair<TreeNode*, int>(tmpPair.first->left, tmpPair.second + 1));
		}
		if (tmpPair.first->right != NULL) {
			treeQ.push(pair<TreeNode*, int>(tmpPair.first->right, tmpPair.second + 1));
		}
	}
}

void testCase1() {
	TreeNode* t1 = new TreeNode(1);
	TreeNode* t2 = new TreeNode(2);
	t1->left = t2;

	Solution s;
	testPrint(s.addOneRow(t1, 100, 1));
	cout << endl;
}

void testCase2() {
	TreeNode* t1 = new TreeNode(1);
	TreeNode* t2 = new TreeNode(2);
	TreeNode* t3 = new TreeNode(3);
	TreeNode* t4 = new TreeNode(4);
	TreeNode* t5 = new TreeNode(5);
	TreeNode* t6 = new TreeNode(6);

	t2->left = t3;
	t2->right = t1;

	t6->left = t5;

	t4->left = t2;
	t4->right = t6;

	Solution s;
	testPrint(s.addOneRow(t4, 1, 2));
	cout << endl;
}

void testCase3() {
	TreeNode* t1 = new TreeNode(1);
	TreeNode* t2 = new TreeNode(2);
	TreeNode* t3 = new TreeNode(3);
	TreeNode* t4 = new TreeNode(4);

	t2->left = t3;
	t2->right = t1;
	t2->left = t3;

	t4->left = t2;

	Solution s;
	testPrint(s.addOneRow(t4, 1, 3));
	cout << endl;
}

int main(void)
{
	testCase1();
	cout << endl;
	testCase2();
	cout << endl;
	testCase3();

	return 0;
}

File: LeetCode/iter1/java/623.add-one-row-to-tree.java
package org.ironman.base.playground.leetcode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Solution {

    public static void main(String[] args) {
        testCase1();
        System.out.println();
        testCase2();
        System.out.println();
        testCase3();
    }

    private static void testCase1() {
        TreeNode t1 = new TreeNode(1);
        TreeNode t2 = new TreeNode(2);
        TreeNode t3 = new TreeNode(3);
        TreeNode t4 = new TreeNode(4);
        TreeNode t5 = new TreeNode(5);
        TreeNode t6 = new TreeNode(6);

        t2.left = t3;
        t2.right = t1;

        t6.left = t5;

        t4.left = t2;
        t4.right = t6;

        printLevelTree(addOneRow(t4, 100, 1));
    }

    private static void testCase2() {
        TreeNode t1 = new TreeNode(1);
        TreeNode t2 = new TreeNode(2);
        TreeNode t3 = new TreeNode(3);
        TreeNode t4 = new TreeNode(4);
        TreeNode t5 = new TreeNode(5);
        TreeNode t6 = new TreeNode(6);

        t2.left = t3;
        t2.right = t1;

        t6.left = t5;

        t4.left = t2;
        t4.right = t6;

        printLevelTree(addOneRow(t4, 1, 2));
    }

    private static void testCase3() {
        TreeNode t1 = new TreeNode(1);
        TreeNode t2 = new TreeNode(2);
        TreeNode t3 = new TreeNode(3);
        TreeNode t4 = new TreeNode(4);

        t2.left = t3;
        t2.right = t1;

        t4.left = t2;

        printLevelTree(addOneRow(t4, 1, 3));
    }

    public static TreeNode addOneRow(TreeNode root, int v, int d) {
        if (d == 1) {
            TreeNode tmp = new TreeNode(v);
            tmp.left = root;
            root = tmp;
            return root;
        }

        Queue<Pair<TreeNode, Integer>> treeNodeQ = new LinkedList<>();
        treeNodeQ.add(new Pair<>(root, 1));

        //find parent level
        List<TreeNode> parentLayer = new ArrayList<>();
        while (!treeNodeQ.isEmpty()) {
            Pair<TreeNode, Integer> tmpPair = treeNodeQ.poll();
            if (tmpPair.getValue() == d - 1) {
                parentLayer.add(tmpPair.key);
            } else {
                if (tmpPair.key.left != null) {
                    treeNodeQ.add(new Pair<>(tmpPair.key.left, tmpPair.value + 1));
                }

                if (tmpPair.key.right != null) {
                    treeNodeQ.add(new Pair<>(tmpPair.key.right, tmpPair.value + 1));
                }
            }
        }

        //add one row to parent layer
        for (TreeNode parentNode : parentLayer) {
            TreeNode leftTmpNode = new TreeNode(v);
            leftTmpNode.left = parentNode.left;
            TreeNode rightTmpNode = new TreeNode(v);
            rightTmpNode.right = parentNode.right;

            parentNode.left = leftTmpNode;
            parentNode.right = rightTmpNode;
        }

        return root;
    }

    private static void printLevelTree(TreeNode root) {
        Queue<Pair<TreeNode, Integer>> q = new LinkedList<>();
        q.add(new Pair<>(root, 1));
        int curLv = 1;
        while (!q.isEmpty()) {
            Pair<TreeNode, Integer> tmpPair = q.poll();
            if (tmpPair.getValue() != curLv) {
                curLv = tmpPair.getValue();
                System.out.println();
            }

            System.out.print(tmpPair.getKey().val + " ");
            if (tmpPair.getKey().left != null) {
                q.add(new Pair<>(tmpPair.getKey().left, tmpPair.getValue() + 1));
            }

            if (tmpPair.getKey().right != null) {
                q.add(new Pair<>(tmpPair.getKey().right, tmpPair.getValue() + 1));
            }
        }
        System.out.println();
    }

    static public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    static public class Pair<K, V> {
        private K key;
        private V value;

        Pair(K _key, V _value) {
            key = _key;
            value = _value;
        }

        public K getKey() {
            return key;
        }

        public V getValue() {
            return value;
        }
    }
}